runtime/sema/check_assignment.go

```diff
@@ -110,6 +113,95 @@ func (checker *Checker) checkAssignment(
 	return
 }
 
+func (checker *Checker) enforceViewAssignment(assignment ast.Statement, target ast.Expression) {
+	if !checker.CurrentPurityScope().EnforcePurity {
+		return
+	}
+
+	isWriteableInViewContext := func(t Type) bool {
+		// We have to prevent any writes to references, since we cannot know where the value
+		// pointed to by the reference may have come from. Similarly, we can never safely assign
+		// to a resource; because resources are moved instead of copied, we cannot currently
+		// track the origin of a write target when it is a resource. Consider:
+		// -----------------------------------------------------------------------------
+		// pub resource R {
+		//   pub(set) var x: Int
+		//   init(x: Int) {
+		//     self.x = x
+		//   }
+		//  }
+		//
+		// view fun foo(_ f: @R): @R {
+		//   let b <- f
+		//   b.x = 3 // b was created in the current scope but modifies the resource value
+		//   return <-b
+		// }
+		_, isReference := t.(*ReferenceType)
+		return !isReference && !t.IsResourceType()
+	}
+
+	var baseVariable *Variable
+	var accessChain []Type = make([]Type, 0)
+	var inAccessChain = true
+
+	// seek the variable expression (if it exists) at the base of the access chain
+	for inAccessChain {
+		switch targetExp := target.(type) {
+		case *ast.IdentifierExpression:
+			baseVariable = checker.valueActivations.Find(targetExp.Identifier.Identifier)
+			if baseVariable != nil {
+				accessChain = append(accessChain, baseVariable.Type)
+			}
+			inAccessChain = false
+		case *ast.IndexExpression:
+			target = targetExp.TargetExpression
+			elementType := checker.visitIndexExpression(targetExp, true)
+			accessChain = append(accessChain, elementType)
+		case *ast.MemberExpression:
+			target = targetExp.Expression
+			memberType, _, _ := checker.visitMember(targetExp)
+			accessChain = append(accessChain, memberType)
+		default:
+			inAccessChain = false
+		}
+	}
+
+	if baseVariable == nil {
```

Comment:
maybe add a brief explanation here about what is meant by `baseVariable == nil`.
I guess it's cases like `foo().something = 5`

Reply:
can we use the type of the expression to perform the check, in such cases? e.g: `foo()`s type is the return type of the function. Target type can be passed by the callee (callee should have this info)

Reply:
It's not just function calls; we also have to consider cases like `b ? s1 : s2`, where `s1` may be local and `s2` might not be.